pkg/exec/devfile_test.go

```diff
@@ -0,0 +1,202 @@
+				},
+			},
+			composite: common.Composite{
+				Id:       commands[2],
+				Commands: []string{commands[0], commands[1]},
+				Parallel: false,
+			},
+			execClient: fakeExecClient,
+			wantErr:    false,
+		},
+		{
+			name: "Case 2: Non-parallel, failed exec",
+			commandsMap: map[string]common.DevfileCommand{
+				commands[0]: {
+					Exec: &common.Exec{Id: commands[0]},
+				},
+				commands[1]: {
+					Exec: &common.Exec{Id: commands[1]},
+				},
+				commands[2]: {
+					Composite: &common.Composite{Id: commands[2]},
+				},
+			},
+			composite: common.Composite{
+				Id:       commands[2],
+				Commands: []string{commands[0], commands[1]},
+				Parallel: false,
+			},
+			execClient: fakeExecErrorClient,
+			wantErr:    true,
+		},
+		{
+			name: "Case 3: Parallel, successful exec",
+			commandsMap: map[string]common.DevfileCommand{
+				commands[0]: {
+					Exec: &common.Exec{Id: commands[0]},
+				},
+				commands[1]: {
+					Exec: &common.Exec{Id: commands[1]},
+				},
+				commands[2]: {
+					Composite: &common.Composite{Id: commands[2]},
+				},
+			},
+			composite: common.Composite{
+				Id:       commands[2],
+				Commands: []string{commands[0], commands[1]},
+				Parallel: true,
+			},
+			execClient: fakeExecClient,
+			wantErr:    false,
+		},
+		{
+			name: "Case 4: Parallel, failed exec",
+			commandsMap: map[string]common.DevfileCommand{
+				commands[0]: {
+					Exec: &common.Exec{Id: commands[0]},
+				},
+				commands[1]: {
+					Exec: &common.Exec{Id: commands[1]},
+				},
+				commands[2]: {
+					Composite: &common.Composite{Id: commands[2]},
+				},
+			},
+			composite: common.Composite{
+				Id:       commands[2],
+				Commands: []string{commands[0], commands[1]},
+				Parallel: true,
+			},
+			execClient: fakeExecErrorClient,
+			wantErr:    true,
+		},
+		{
+			name: "Case 5: Non-Parallel, command not found",
+			commandsMap: map[string]common.DevfileCommand{
+				commands[0]: {
+					Exec: &common.Exec{Id: commands[0]},
+				},
+				commands[1]: {
+					Exec: &common.Exec{Id: commands[1]},
+				},
+				commands[2]: {
+					Composite: &common.Composite{Id: commands[2]},
+				},
+			},
+			composite: common.Composite{
+				Id:       commands[2],
+				Commands: []string{commands[0], "fake-command"},
+				Parallel: false,
+			},
+			execClient: fakeExecClient,
+			wantErr:    true,
+		},
+		{
+			name: "Case 6: Parallel, command not found",
+			commandsMap: map[string]common.DevfileCommand{
+				commands[0]: {
+					Exec: &common.Exec{Id: commands[0]},
+				},
+				commands[1]: {
+					Exec: &common.Exec{Id: commands[1]},
+				},
+				commands[2]: {
+					Composite: &common.Composite{Id: commands[2]},
+				},
+			},
+			composite: common.Composite{
+				Id:       commands[2],
+				Commands: []string{commands[0], "fake-command"},
+				Parallel: true,
+			},
+			execClient: fakeExecClient,
+			wantErr:    true,
+		},
+		{
+			name: "Case 7: Nested composite commands",
```

Comment:
probably add another case for nested parallel composite commands

Reply:
Done 👍